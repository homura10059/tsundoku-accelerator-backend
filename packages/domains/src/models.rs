use crate::item_metadata::ItemMetaData;
use anyhow::Result;
use url::Url;

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct WishListSnapshot {
    pub id: String,
    pub title: String,
    pub url: Url,
    pub scraped_at: i64,
    pub items: Vec<ItemMetaData>,
}
