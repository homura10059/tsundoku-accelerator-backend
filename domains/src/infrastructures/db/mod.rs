pub mod ebook_snapshot;
pub mod prisma;
pub mod wish_list;

use anyhow::Result;
use prisma::PrismaClient;

pub async fn get_client() -> Result<PrismaClient> {
    let client = prisma::new_client().await?;
    Ok(client)
}

#[cfg(test)]
mod tests {
    use super::*;
    use dotenv;

    #[tokio::test]
    async fn test_get_client() {
        dotenv::dotenv().ok();

        let client = get_client().await;
    }
}
