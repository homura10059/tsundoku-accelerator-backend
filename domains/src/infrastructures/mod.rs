pub mod db;
pub mod prisma;
pub mod scraper;
