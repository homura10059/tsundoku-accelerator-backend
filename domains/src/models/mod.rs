use url::Url;

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct ItemMetaData {
    pub id: String,
    pub url: Url,
    pub price: String,
}

#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, Clone)]
pub struct WishListSnapshot {
    pub id: String,
    pub title: String,
    pub url: Url,
    pub scraped_at: i64,
    pub items: Vec<ItemMetaData>,
}
