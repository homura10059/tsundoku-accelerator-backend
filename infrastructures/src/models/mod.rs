use url::Url;

#[derive(Debug, PartialEq, Clone)]
pub struct ItemMetaData {
    pub id: String,
    pub url: Url,
}

#[derive(Debug, PartialEq, Clone)]
pub struct WishListSnapshot {
    pub id: String,
    pub title: String,
    pub url: Url,
    pub scraped_at: i64,
    pub items: Vec<ItemMetaData>,
}
