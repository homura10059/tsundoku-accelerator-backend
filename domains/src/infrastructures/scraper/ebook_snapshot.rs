use crate::infrastructures::scraper;
use crate::models::{EbookSnapshot, Payment};
use anyhow::Result;
use headless_chrome::Browser;
use once_cell::sync::Lazy;
use pure_funcs::get_now_in_sec;
use regex::Regex;
use std::collections::HashSet;
use url::Url;

static PRICE_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"￥(?P<price>\d{1,3}(,\d{3})*)").unwrap());
static POINT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?P<points>\d{1,3}(,\d{3})*)pt").unwrap());

pub fn create_url(id: &str) -> Result<Url> {
    let url = Url::parse("https://www.amazon.co.jp/dp/")?;
    let joined = url.join(id)?;
    Ok(joined)
}

fn extract_price(text: &str) -> HashSet<&str> {
    let matched = PRICE_REGEX
        .captures_iter(text)
        .map(|cap| cap.name("price").unwrap().as_str())
        .collect::<Vec<_>>();
    matched.into_iter().collect()
}

fn extract_points(text: &str) -> HashSet<&str> {
    let matched = POINT_REGEX
        .captures_iter(text)
        .map(|cap| cap.name("points").unwrap().as_str())
        .collect::<Vec<_>>();
    matched.into_iter().collect()
}

fn extract_payment(text: &str) -> Option<Payment> {
    let price = *extract_price(text).iter().next()?;
    let points = *extract_points(text).iter().next()?;
    Some(Payment {
        price: price.to_string(),
        points: points.to_string(),
    })
}

pub fn get(id: &str) -> Result<EbookSnapshot> {
    let url = create_url(id)?;
    let browser = Browser::default()?;

    let tab = browser.wait_for_initial_tab()?;
    tab.navigate_to(url.as_str())?;
    tab.wait_for_element("#navFooter")?;

    let title_element = tab.find_element("#productTitle")?;
    let title = title_element.get_inner_text()?;

    let image = tab.find_element("#ebooksImgBlkFront")?;
    let image_attribute = image.get_attributes()?.unwrap();
    let thumbnail_url_str = scraper::search_from(&image_attribute, "src").unwrap();
    let thumbnail_url = Url::parse(thumbnail_url_str.as_str())?;

    let price_list = tab.find_elements("#tmmSwatches .a-button-text")?;
    let payments = price_list
        .iter()
        .flat_map(|price| price.get_inner_text())
        .flat_map(|text| {
            let payment = extract_payment(text.as_str());
            payment
        })
        .collect::<Vec<_>>();

    let snapshot = EbookSnapshot {
        ebook_id: id.to_string(),
        scraped_at: get_now_in_sec(),
        title: String::from(title.trim()),
        thumbnail_url,
        payment_ebook: payments.get(0).map(|x| x.clone()),
        payment_real: payments.get(1).map(|x| x.clone()),
    };

    Ok(snapshot)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_get() {
        let id = String::from("B09RQGMYKZ");
        let actual = get(id.as_str()).unwrap();
        assert_eq!(actual.ebook_id, id);
        assert_eq!(actual.title, String::from("コンセプトから理解するRust"));
        assert_eq!(
            actual.thumbnail_url.to_string(),
            String::from("https://m.media-amazon.com/images/I/51CTnaTcJtL.jpg")
        );
        assert_debug_snapshot!(actual.payment_ebook);
        assert_debug_snapshot!(actual.payment_real);
    }

    #[test]
    fn test_extract_payment() {
        let actual = extract_payment(
            r#"
            Kindle版 (電子書籍)
            ￥3,344
            獲得ポイント: 152pt
            "#,
        );
        let expected = Some(Payment {
            price: String::from("3,344"),
            points: String::from("152"),
        });
        assert_eq!(actual, expected);
    }
}
