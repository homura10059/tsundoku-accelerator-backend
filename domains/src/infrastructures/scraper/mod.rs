mod item_meta_data;
mod wish_list;

use crate::models::ItemMetaData;
use crate::models::WishListSnapshot;
use anyhow::Result;
use headless_chrome::Browser;
use pure_funcs::get_now_in_sec;

pub fn get_wish_list_snapshot(id: &str) -> Result<WishListSnapshot> {
    let url = wish_list::create_url(id)?;
    let browser = Browser::default()?;

    let tab = browser.wait_for_initial_tab()?;
    tab.navigate_to(url.as_str())?;

    let nav_to_top = tab.wait_for_element("#navBackToTop")?;

    while tab.find_element("#endOfListMarker").is_err() {
        nav_to_top.scroll_into_view()?;
        break;
    }

    let links = tab.find_elements(".a-link-normal")?;

    let mut items: Vec<_> = links
        .iter()
        .filter_map(|link| link.get_attributes().ok())
        .filter_map(|x| x)
        .flatten()
        .filter(|text| text.contains("/dp/"))
        .filter_map(|href| item_meta_data::create(href).ok())
        .collect();
    items.sort();
    items.dedup();

    let title_element = tab.find_element("#profile-list-name")?;
    let title = title_element.get_inner_text()?;

    let snapshot = WishListSnapshot {
        id: id.to_string(),
        url,
        scraped_at: get_now_in_sec(),
        title,
        items,
    };

    Ok(snapshot)
}

#[cfg(test)]
mod tests {
    use super::*;
    use url::Url;

    fn get_items_mock() -> Vec<ItemMetaData> {
        let mut items = vec![
            ItemMetaData {
                id: String::from("B08S7CJV4X"),
                url: Url::parse("https://www.amazon.co.jp/dp/B08S7CJV4X/").unwrap(),
            },
            ItemMetaData {
                id: String::from("B08L53NT5P"),
                url: Url::parse("https://www.amazon.co.jp/dp/B08L53NT5P/").unwrap(),
            },
            ItemMetaData {
                id: String::from("B08L54335M"),
                url: Url::parse("https://www.amazon.co.jp/dp/B08L54335M/").unwrap(),
            },
            ItemMetaData {
                id: String::from("B08L51YSLR"),
                url: Url::parse("https://www.amazon.co.jp/dp/B08L51YSLR/").unwrap(),
            },
            ItemMetaData {
                id: String::from("B08L5278XF"),
                url: Url::parse("https://www.amazon.co.jp/dp/B08L5278XF/").unwrap(),
            },
            ItemMetaData {
                id: String::from("B09TPLQGKS"),
                url: Url::parse("https://www.amazon.co.jp/dp/B09TPLQGKS/").unwrap(),
            },
            ItemMetaData {
                id: String::from("B0B5Q2RMX6"),
                url: Url::parse("https://www.amazon.co.jp/dp/B0B5Q2RMX6/").unwrap(),
            },
        ];
        items.sort();
        items
    }

    #[test]
    fn test_get_item_urls() {
        let items = get_items_mock();
        let expected = WishListSnapshot {
            id: String::from("2BDAPI9RQ09E9"),
            url: Url::parse("https://www.amazon.jp/hz/wishlist/ls/2BDAPI9RQ09E9").unwrap(),
            title: String::from("do_not_delete"),
            scraped_at: 0,
            items,
        };
        let actual = get_wish_list_snapshot(expected.id.as_str()).unwrap();
        assert_eq!(actual.id, expected.id);
        assert_eq!(actual.url, expected.url);
        assert_eq!(actual.title, expected.title);
        assert_eq!(actual.items, expected.items);
    }
}
