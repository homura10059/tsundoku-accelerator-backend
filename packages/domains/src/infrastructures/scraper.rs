use crate::item_metadata::ItemMetaData;
use crate::models::WishListSnapshot;
use anyhow::Result;
use headless_chrome::{Browser, Element};
use pure_funcs::get_now_in_sec;
use url::Url;

pub fn search_from(attributes: &Vec<String>, key: &str) -> Option<String> {
    let target = attributes.iter().position(|attr| attr == key)? + 1;
    let (_pos, attr) = attributes
        .iter()
        .enumerate()
        .find(|(pos, _attr)| pos.eq(&target))?;
    let result = attr.clone();
    Some(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn search_from_returns_some() {
        let mock_vec = vec!["key_1".to_string(), "val_1".to_string()];
        let actual = search_from(&mock_vec, "key_1");
        assert_eq!(actual, Some("val_1".to_string()))
    }

    #[test]
    fn search_from_returns_none() {
        let mock_vec = vec!["key_1".to_string(), "val_1".to_string()];
        let actual = search_from(&mock_vec, "key_999");
        assert_eq!(actual, None)
    }

    #[test]
    fn search_from_returns_none2() {
        let mock_vec = vec!["key_1".to_string(), "val_1".to_string()];
        let actual = search_from(&mock_vec, "val_1");
        assert_eq!(actual, None)
    }
}
