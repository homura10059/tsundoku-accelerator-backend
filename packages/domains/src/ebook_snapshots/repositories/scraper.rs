use crate::ebook_snapshots::model::{EbookSnapshot, Payment};
use anyhow::Result;
use chrono::Utc;
use headless_chrome::Browser;
use infrastructures::scraper;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashSet;
use url::Url;

static PRICE_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"￥(?P<price>\d{1,3}(,\d{3})*)").unwrap());
static POINT_REGEX: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?P<points>\d{1,3}(,\d{3})*)pt").unwrap());

pub fn create_url(id: &str) -> Result<Url> {
    let url = Url::parse("https://www.amazon.co.jp/dp/")?;
    let joined = url.join(id)?;
    Ok(joined)
}

fn extract_price(text: &str) -> HashSet<&str> {
    let matched = PRICE_REGEX
        .captures_iter(text)
        .map(|cap| cap.name("price").unwrap().as_str())
        .collect::<Vec<_>>();
    matched.into_iter().collect()
}

fn extract_points(text: &str) -> HashSet<&str> {
    let matched = POINT_REGEX
        .captures_iter(text)
        .map(|cap| cap.name("points").unwrap().as_str())
        .collect::<Vec<_>>();
    matched.into_iter().collect()
}

fn extract_payment(text: &str) -> Option<Payment> {
    let price = *extract_price(text).iter().next()?;
    let points = *extract_points(text).iter().next()?;
    Some(Payment {
        price: price.to_string().trim().replace(',', ""),
        points: points.to_string().trim().replace(',', ""),
    })
}

pub fn get(browser: &Browser, id: &str) -> Result<EbookSnapshot> {
    let url = create_url(id)?;

    let tab = browser.new_tab()?;
    tab.navigate_to(url.as_str())?;
    tab.wait_for_element("#navFooter")?;

    let image = tab.find_element("#ebooksImgBlkFront")?;
    let image_attribute = image.get_attributes()?.unwrap();
    let thumbnail_url_str = scraper::search_from(&image_attribute, "src").unwrap();
    let thumbnail_url = Url::parse(thumbnail_url_str.as_str())?;

    let price_list = tab.find_elements("#tmmSwatches .a-button-text")?;
    let payments = price_list
        .iter()
        .flat_map(|price| price.get_inner_text())
        .flat_map(|text| {
            let payment = extract_payment(text.as_str());
            payment
        })
        .collect::<Vec<_>>();

    let snapshot = EbookSnapshot {
        ebook_id: id.to_string(),
        scraped_at: Utc::now().timestamp(),
        thumbnail_url,
        payment_ebook: payments.get(0).cloned(),
        payment_real: payments.get(1).cloned(),
    };

    tab.close(true)?;

    Ok(snapshot)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_get() {
        let browser = Browser::default().unwrap();
        let id = String::from("B09RQGMYKZ");
        let actual = get(&browser, id.as_str()).unwrap();
        assert_eq!(actual.ebook_id, id);
        assert_debug_snapshot!(actual.payment_ebook);
        assert_debug_snapshot!(actual.payment_real);
    }

    #[test]
    fn test_extract_payment() {
        let actual = extract_payment(
            r#"
            Kindle版 (電子書籍)
            ￥3,344
            獲得ポイント: 152pt
            "#,
        );
        let expected = Some(Payment {
            price: String::from("3344"),
            points: String::from("152"),
        });
        assert_eq!(actual, expected);
    }
}
