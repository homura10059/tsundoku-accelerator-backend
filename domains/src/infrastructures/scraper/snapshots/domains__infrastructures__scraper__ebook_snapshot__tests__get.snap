---
source: domains/src/infrastructures/scraper/ebook_snapshot.rs
expression: actual.payment_ebook
---
Some(
    Payment {
        price: "3344",
        points: "152",
    },
)
